test/test_tools.py

```diff
@@ -5,8 +5,6 @@
 from datetime import timedelta
 import re
 
-import pytest
```

Comment:
Won't we get a drop in coverage by removing these tests while the tested functions still exist? Coveralls reported -0.1% for the current patch. I think deleting these tests is premature.

I also think it's important to make sure the equivalent behavior is tested in `plugins.rules` and its various classes. Obviously `<RuleTypeClass>.get_rule_regex()` must be called in the course of setting up the tests in `test/plugins/test_plugins_rules.py`, so its behavior in each class is validated—but there's e.g. minimal validation of command rules' match groups (AFAICT, only groups 0 & 1).

Reply:
I consider this:

* deprecated functions **must not** be modified once they are deprecated
* these functions have been tested before deprecation without issues

So I was a bit quick to delete these tests, I admit.

However, about the actual tests: https://github.com/sopel-irc/sopel/blob/b16cfcd8e1197e4e85727f00f56c69f01c890aac/test/plugins/test_plugins_rules.py#L1421-L1501

So it partially cover this, because only the commands are, not the nickname/action commands.

Reply:
Honestly, though, these "deprecated functions" aren't actually deprecated until this PR gets merged, so you are still allowed to change them. 😝 But my main concern is behavior drift, making sure the deprecated functions _and_ the new stuff in `plugins.rules` are doing the same things. Plus, uncovered lines make me sad. (_You_ did this to me!)

It's possible that I'm being overly paranoid, but heh. Your job is being a test-writing wizard, and my job is fretting about backward compatibility to ridiculous extremes. 😁

Hmm, there _are_ a few places where match groups other than 0 & 1 get checked for action & nick commands, but they're asserted to be `None`. We don't have _positive_ tests for groups in those command types to check that they get set correctly, only tests that verify they _aren't_ set when not expected.

The `rules` test stuff is enough to be a "separate PR" kind of thing now, so I'll leave it alone. But I'll reserve the right to take a stab at turning the deprecated functions into wrappers around the new stuff.

Reply:
Update: Pretty simple to turn the pattern functions into wrappers, actually. I should do the same for the regexp ones.